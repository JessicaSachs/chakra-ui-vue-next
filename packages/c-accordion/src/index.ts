import { h, defineComponent, PropType } from 'vue'
import { chakra, DOMElements } from '@chakra-ui/vue-system'

const CAccordion = defineComponent({
  props: {
    as: {
      type: [String] as PropType<DOMElements>,
      default: 'div',
    },
  },
  setup(props, { slots, attrs }) {
    return () => h(chakra(props.as), { ...attrs }, slots)
  },
})

export default CAccordion
