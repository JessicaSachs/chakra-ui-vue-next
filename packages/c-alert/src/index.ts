import { h, defineComponent, Component } from 'vue'

const CAlert = defineComponent({
  props: {
    as: {
      type: [String, Object],
      default: 'div',
    },
  },
  render() {
    return h(
      this.as,
      {
        ...this.$attrs,
        role: 'alert',
      },
      this.$slots.default && this.$slots.default()
    )
  },
})

export default CAlert
