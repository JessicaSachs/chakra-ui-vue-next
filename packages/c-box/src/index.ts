import { h, defineComponent, PropType } from 'vue'

const CBox = defineComponent({
  props: {
    as: {
      type: String as PropType<string>,
      default: 'div',
    },
  },
  render() {
    return h(this?.as, { ...this.$props, ...this.$attrs }, this.$slots)
  },
})

export default CBox
