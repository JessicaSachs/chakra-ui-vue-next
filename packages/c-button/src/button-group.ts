import { computed, defineComponent, h, PropType } from 'vue'
import {
  SystemCSSProperties,
  SystemStyleObject,
} from '@chakra-ui/styled-system'
import { chakra, ThemingProps } from '@chakra-ui/vue-system'
import { ComponentThemeConfig } from '@chakra-ui/vue-theme'
import { createContext } from '@chakra-ui/vue-utils'

export interface ButtonGroupProps extends ThemingProps {
  /**
   * If `true`, the borderRadius of button that are direct children will be altered
   * to look flushed together
   */
  isAttached?: boolean
  /**
   * If `true`, all wrapped button will be disabled
   */
  isDisabled?: boolean
  /**
   * The spacing between the buttons
   * @default '0.5rem'
   * @type SystemProps["marginRight"]
   */
  spacing?: SystemCSSProperties['marginRight']
}

const props = {
  isAttached: Boolean as PropType<ButtonGroupProps['isAttached']>,
  isDisabled: Boolean as PropType<ButtonGroupProps['isDisabled']>,
  spacing: [String, Number, Array] as PropType<ButtonGroupProps['spacing']>,
  variant: {
    type: String as PropType<string>,
    default: 'solid',
  },
  size: {
    type: String as PropType<string>,
    default: 'sm',
  },
  colorScheme: String as PropType<string>,
  styleConfig: String as PropType<ComponentThemeConfig>,
}

interface ButtonGroupContext extends ThemingProps {
  isDisabled?: boolean
}

const [ButtonGroupProvider, useButtonGroup] = createContext<ButtonGroupContext>(
  {
    strict: false,
    name: 'ButtonGroupContext',
  }
)

export const CButtonGroup = defineComponent({
  name: 'CButtonGroup',
  props,
  setup(props, { slots }) {
    ButtonGroupProvider({
      size: props.size,
      colorScheme: props.colorScheme,
      variant: props.variant,
      isDisabled: props.isDisabled,
    })

    const styles = computed(() => {
      let groupStyles: SystemStyleObject = {
        display: 'inline-flex',
      }

      if (props.isAttached) {
        groupStyles = {
          ...groupStyles,
          '> *:first-of-type:not(:last-of-type)': { borderRightRadius: 0 },
          '> *:not(:first-of-type):not(:last-of-type)': { borderRadius: 0 },
          '> *:not(:first-of-type):last-of-type': { borderLeftRadius: 0 },
        }
      } else {
        groupStyles = {
          ...groupStyles,
          '& > *:not(style) ~ *:not(style)': { marginLeft: props.spacing },
        }
      }

      return groupStyles
    })

    return () => h(chakra.div)
  },
})
