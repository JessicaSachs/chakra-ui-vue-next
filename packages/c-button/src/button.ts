import { h, defineComponent, PropType, computed } from 'vue'
import {
  chakra,
  DOMElements,
  useStyleConfig,
  ThemingProps,
} from '@chakra-ui/vue-system'
import {
  SystemCSSProperties,
  SystemStyleObject,
} from '@chakra-ui/styled-system'
import { ComponentThemeConfig } from '@chakra-ui/vue-theme'
import { dataAttr } from '@chakra-ui/vue-utils'

type ButtonTypes = 'button' | 'reset' | 'submit'

const props = {
  as: {
    type: String as PropType<DOMElements>,
    default: 'button',
  },
  isLoading: Boolean as PropType<boolean>,
  isActive: Boolean as PropType<boolean>,
  isDisabled: Boolean as PropType<boolean>,
  loadingText: String as PropType<string>,
  isFullWidth: Boolean as PropType<boolean>,
  type: String as PropType<ButtonTypes>,
  leftIcon: String as PropType<string>,
  rightIcon: String as PropType<string>,
  colorScheme: String as PropType<string>,
  variant: {
    type: String as PropType<string>,
    default: 'solid',
  },
  size: {
    type: String as PropType<string>,
    default: 'md',
  },
  styleConfig: String as PropType<ComponentThemeConfig>,

  /** Not sure if the SystemCSSProperties is the right prop type for this */
  iconSpacing: {
    type: [String, Number, Array] as PropType<
      SystemCSSProperties['marginRight']
    >,
    default: '0.5rem',
  },
}

/** TODO: How to get component props in typescript */

const CButton = defineComponent({
  name: 'CButton',
  props,
  setup(props, { attrs, slots }) {
    const themingProps = computed<ThemingProps>(() => ({
      colorScheme: props.colorScheme,
      variant: props.variant,
      size: props.size,
      styleConfig: props.styleConfig,
    }))

    const styles = useStyleConfig('Button', themingProps.value)

    const buttonStyles: SystemStyleObject = {
      display: 'inline-flex',
      appearance: 'none',
      alignItems: 'center',
      justifyContent: 'center',
      transition: 'all 250ms',
      userSelect: 'none',
      position: 'relative',
      whiteSpace: 'nowrap',
      verticalAlign: 'middle',
      outline: 'none',
      width: props.isFullWidth ? '100%' : 'auto',
      ...styles.value,
    }

    return () =>
      h(
        chakra(props.as, 'button'),
        {
          disabled: props.isDisabled || props.isLoading,
          type: props.as === 'button' ? undefined : props.type,
          dataActive: dataAttr(props.isActive),
          dataLoading: dataAttr(props.isLoading),
          ...buttonStyles,
          ...attrs,
        },
        slots
      )
  },
})

export default CButton
