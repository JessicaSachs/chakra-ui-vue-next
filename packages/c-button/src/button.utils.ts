import { SystemCSSProperties } from '@chakra-ui/styled-system'
import { DOMElements, ThemingProps } from '@chakra-ui/vue-system'
import { PropType } from 'vue'

type ButtonTypes = 'button' | 'reset' | 'submit'

export const BUTTON_PROPS = {
  as: {
    type: String as PropType<DOMElements>,
    default: 'button',
  },
  isLoading: Boolean as PropType<boolean>,
  isActive: Boolean as PropType<boolean>,
  isDisabled: Boolean as PropType<boolean>,
  loadingText: String as PropType<string>,
  isFullWidth: Boolean as PropType<boolean>,
  type: String as PropType<ButtonTypes>,
  leftIcon: String as PropType<string>,
  rightIcon: String as PropType<string>,
  colorScheme: String as PropType<ThemingProps['colorScheme']>,
  variant: String as PropType<ThemingProps['variant']>,
  size: String as PropType<ThemingProps['size']>,
  styleConfig: String as PropType<ThemingProps['styleConfig']>,

  /** Not sure if the SystemCSSProperties is the right prop type for this */
  iconSpacing: {
    type: [String, Number, Array] as PropType<
      SystemCSSProperties['marginRight']
    >,
    default: '0.5rem',
  },
}
