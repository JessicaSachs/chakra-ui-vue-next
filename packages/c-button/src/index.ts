import { h, defineComponent, PropType } from 'vue'

const CButton = defineComponent({
  name: 'CButton',
  props: {
    as: {
      type: String as PropType<string>,
      default: 'button',
    },
  },
  render() {
    return h(
      this?.as,
      { ...this.$props, ...this.$attrs },
      this.$slots.default && this.$slots.default()
    )
  },
})

export default CButton
