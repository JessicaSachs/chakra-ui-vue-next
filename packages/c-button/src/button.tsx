import { h, defineComponent, PropType, computed, cloneVNode } from 'vue'
import {
  chakra,
  DOMElements,
  useStyleConfig,
  ThemingProps,
} from '@chakra-ui/vue-system'
import {
  SystemCSSProperties,
  SystemStyleObject,
} from '@chakra-ui/styled-system'
import { ComponentThemeConfig } from '@chakra-ui/vue-theme'
import { dataAttr, mergeWith } from '@chakra-ui/vue-utils'
import { useButtonGroup } from './button-group'

type ButtonTypes = 'button' | 'reset' | 'submit'

const props = {
  as: {
    type: String as PropType<DOMElements>,
    default: 'button',
  },
  isLoading: Boolean as PropType<boolean>,
  isActive: Boolean as PropType<boolean>,
  isDisabled: Boolean as PropType<boolean>,
  loadingText: String as PropType<string>,
  isFullWidth: Boolean as PropType<boolean>,
  type: String as PropType<ButtonTypes>,
  leftIcon: String as PropType<string>,
  rightIcon: String as PropType<string>,
  colorScheme: String as PropType<string>,
  variant: {
    type: String as PropType<string>,
    default: 'solid',
  },
  size: {
    type: String as PropType<string>,
    default: 'md',
  },
  styleConfig: String as PropType<ComponentThemeConfig>,

  /** Not sure if the SystemCSSProperties is the right prop type for this */
  iconSpacing: {
    type: [String, Number, Array] as PropType<
      SystemCSSProperties['marginRight']
    >,
    default: '0.5rem',
  },
}

/**
 * CButton
 * 
 * The Button component is used to trigger an action or event,
 * such as submitting a form, opening a dialog, canceling
 * an action, or performing a delete operation.
 */
const CButton = defineComponent({
  name: 'CButton',
  props,
  setup(props, { attrs, slots }) {
    const themingProps = computed<ThemingProps>(() => ({
      colorScheme: props.colorScheme,
      variant: props.variant,
      size: props.size,
      styleConfig: props.styleConfig,
    }))

    const group = useButtonGroup()
    const styles = useStyleConfig('Button', { ...group, ...themingProps.value })

    const _focus = mergeWith({}, styles.value?.['_focus'] ?? {}, { zIndex: 1 })

    const buttonStyles: SystemStyleObject = {
      display: 'inline-flex',
      appearance: 'none',
      alignItems: 'center',
      justifyContent: 'center',
      transition: 'all 250ms',
      userSelect: 'none',
      position: 'relative',
      whiteSpace: 'nowrap',
      verticalAlign: 'middle',
      outline: 'none',
      width: props.isFullWidth ? '100%' : 'auto',
      ...styles.value,
      ...(!!group && { _focus }),
    }

    return () => (
      <chakra.button
      // @ts-ignore JSX props error
        as={props.as}
        label="button"
        disabled={props.isDisabled || props.isLoading}
        type={props.as === 'button' ? undefined : props.type}
        dataActive={dataAttr(props.isActive)}
        dataLoading={dataAttr(props.isLoading)}
        {...buttonStyles}
        {...attrs}
      >
        {slots}
      </chakra.button>
    )
  },
})

const CButtonSpinner = defineComponent({
  name: 'CButtonSpinner',
  props: {
    label: Boolean as PropType<boolean>,
    spacing: [Number, String, Array] as PropType<
      number | string | string[] | number[]
    >,
    __css: Object as PropType<SystemStyleObject>,
  },
  setup(props, { attrs, slots }) {

    const spinnerStyles: SystemStyleObject = {
      display: "flex",
      alignItems: "center",
      position: props.label ? "relative" : "absolute",
      marginEnd: props.label ? props.spacing : 0,
      ...props.__css,
    }
    return () => (
      // @ts-ignore JSX props error
      <chakra.div label="button__spinner" baseStyle={{}} {...attrs} __css={spinnerStyles}>
        {slots ?? slots}
      </chakra.div>
    )
  },
})

/**
 * CButtonIcon
 * 
 * Button icon component
 */
const CButtonIcon = defineComponent({
  name: 'CButtonIcon',
  setup(_, { slots, attrs }) {
    const children = slots?.default?.()
    const _children = children
      ? cloneVNode(children?.[0], {
          'aria-hidden': true,
          focusable: false,
        })
      : children?.[0]

    return () =>
      <chakra.span label="button__icon" {...attrs}>
        {_children}
      </chakra.span>
  },
})

export default CButton
