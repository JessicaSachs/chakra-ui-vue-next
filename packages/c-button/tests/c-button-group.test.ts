import { render } from '../../test-utils/src'
import { CButtonGroup, CButton } from '../src'

const renderComponent = (props?: any) => {
  const base = {
    components: {
      CButtonGroup,
      CButton,
    },
    template: `
    <c-button-group variant="outline">
      <c-button color-scheme="blue">Save</c-button>
      <c-button>Cancel</c-button>
    </c-button-group>
    `,
    ...props,
  }
  return render(base)
}

it('should render properly', () => {
  const { asFragment } = renderComponent()
  expect(asFragment()).toMatchSnapshot()
})
