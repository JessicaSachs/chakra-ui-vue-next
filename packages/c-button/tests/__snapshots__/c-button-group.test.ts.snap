// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<DocumentFragment>
  <div
    class="chakra-button__group css-v5zn0n"
    role="group"
  >
    <button
      class="chakra-button css-1bczd7i"
    >
      <!---->
      <!---->
      Save
      <!---->
    </button>
    <button
      class="chakra-button css-1rwqev2"
    >
      <!---->
      <!---->
      Cancel
      <!---->
    </button>
  </div>
</DocumentFragment>
`;
